orangecanvas/help/provider.py

```diff
@@ -47,49 +71,53 @@ def __init__(self, inventory, parent=None):
-
-        try:
-            os.makedirs(cache_dir, exist_ok=True)
-        except OSError:
-            pass
-
-        url = QUrl(self.inventory)
+    def _fetch_inventory(self, url: QUrl) -> None:
         if not url.isLocalFile():
             # fetch and cache the inventory file.
-            manager = QNetworkAccessManager(self)
-            cache = QNetworkDiskCache()
-            cache.setCacheDirectory(cache_dir)
-            manager.setCache(cache)
+            self._manager = manager = self._networkAccessManagerInstance()
             req = QNetworkRequest(url)
-
-            # Follow redirects (for example http -> https)
-            # If redirects were not followed, the documentation would not be found
-            try:
-                req.setAttribute(QNetworkRequest.FollowRedirectsAttribute, 1)  # from Qt 5.6
-                req.setAttribute(QNetworkRequest.RedirectPolicyAttribute,  # from Qt 5.9
-                                 QNetworkRequest.NoLessSafeRedirectPolicy)
-            except AttributeError:  # if ran with earlier Qt
-                pass
-
+            req.setAttribute(
+                QNetworkRequest.CacheLoadControlAttribute,
+                QNetworkRequest.PreferCache
+            )
+            req.setAttribute(
+                QNetworkRequest.FollowRedirectsAttribute, True
+            )
+            req.setAttribute(
+                QNetworkRequest.RedirectPolicyAttribute,
+                QNetworkRequest.NoLessSafeRedirectPolicy
+            )
+            req.setMaximumRedirectsAllowed(5)
             self._reply = manager.get(req)
-            manager.finished.connect(self._on_finished)
+            self._reply.finished.connect(self._on_finished)
         else:
             with open(url.toLocalFile(), "rb") as f:
                 self._load_inventory(f)
 
-    def _on_finished(self, reply):
-        # type: (QNetworkReply) -> None
+    @pyqtSlot()
+    def _on_finished(self):
+        # type: () -> None
+        assert self._reply.isFinished()
+        assert self.sender() is self._reply
+        reply = self._reply  # type: QNetworkReply
+        if log.level <= logging.DEBUG:
+            s = io.StringIO()
+            print("\nGET:", reply.url().toString(), file=s)
+            if reply.attribute(QNetworkRequest.SourceIsFromCacheAttribute):
+                print("  (served from cache)", file=s)
+            for name, val in reply.rawHeaderPairs():
+                print(bytes(name).decode("latin-1"), ":",
+                      bytes(val).decode("latin-1"), file=s)
+            log.debug(s.getvalue())
         if reply.error() != QNetworkReply.NoError:
             log.error("An error occurred while fetching "
                       "help inventory '{0}'".format(self.inventory))
             self._error = reply.error(), reply.errorString()
-
         else:
             contents = bytes(reply.readAll())
             self._load_inventory(io.BytesIO(contents))
+        self._reply = None
```

Comment:
This is probably the reason I get the following crash on line 275 (when no local docs):
```
Traceback (most recent call last):
  File "/home/lan/dev/orange-canvas-core/orangecanvas/application/canvasmain.py", line 1937, in event
    url = self.help.search(url)
  File "/home/lan/dev/orange-canvas-core/orangecanvas/help/manager.py", line 121, in search
    return provider.search(desc)
  File "/home/lan/dev/orange-canvas-core/orangecanvas/help/provider.py", line 275, in search
    if not self.inventory.isLocalFile() and not self._reply.isFinished():
AttributeError: 'NoneType' object has no attribute 'isFinished'
```